### PR TITLE
Hy_invs'

### PR DESCRIPTION
**Pull request by mktnk3**
This PR proves `hy_invs’` and `handleRecv_isBlocking_corres’` (-2 sorries).

Also:
- removes an unnecessary precondition from receiveIPC_corres
- tweaks `handle_recv` and `handleRecv` spec

### CODE REVIEW

**Comment by mktnk3** [comment]
The failures with the proofs are something to do with the links not working (for HaskellKernel specifically) at the moment. Will try again later.

File: proof/invariant-abstract/DetSchedDomainTime_AI.thy [hunk 1]
```diff
@@ -305,9 +305,8 @@ crunch domain_list_inv[wp]: handle_invocation,receive_ipc,receive_signal "\<lamb
 lemma handle_recv_domain_list_inv[wp]:
   "\<lbrace>\<lambda>s::det_state. P (domain_list s)\<rbrace>
   handle_recv is_blocking can_reply \<lbrace>\<lambda>rv s. P (domain_list s)\<rbrace>"
-  apply (wpsimp simp: handle_recv_def Let_def whenE_def get_sk_obj_ref_def
+  by (wpsimp simp: handle_recv_def Let_def whenE_def get_sk_obj_ref_def
```
Comment (michaelmcinerney) [inline proof/invariant-abstract/DetSchedDomainTime_AI.thy:308]: Indentation

File: proof/refine/ARM/Syscall_R.thy [hunk 2]
```diff
@@ -1522,72 +1522,181 @@ crunch sch_act_sane[wp]: cteDeleteOne sch_act_sane
...
-  apply (simp add: handle_recv_def handleRecv_def liftM_bind Let_def
-                   cap_register_def capRegister_def
-             cong: if_cong cap.case_cong capability.case_cong bool.case_cong split del: if_split)
-  sorry (*
+  unfolding handle_recv_def handleRecv_def Let_def
+  apply add_cur_tcb'
+  apply (rule_tac Q="ct_active'" in corres_cross_add_guard)
+   apply (fastforce elim!: ct_active_cross dest: invs_psp_aligned invs_distinct)
+  apply (simp add: cap_register_def capRegister_def liftM_bind
+            cong: if_cong cap.case_cong capability.case_cong split del: if_split)
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Syscall_R.thy:1617]: Indentation

File: proof/refine/ARM/Syscall_R.thy [hunk 3]
```diff
@@ -1522,72 +1522,181 @@ crunch sch_act_sane[wp]: cteDeleteOne sch_act_sane
...
+         apply (case_tac epCap; simp split del: if_split)
+                      apply wpsimp
+                     apply wpsimp
+                    apply (rename_tac readright; case_tac readright; simp)
+                     apply wp
+                       apply simp
+                       apply wp
+                      apply (wp getNotification_wp)
+                     apply clarsimp
+                    apply (wpsimp+)[13]
```
Comment (michaelmcinerney) [inline proof/refine/ARM/Syscall_R.thy:1686]: Is there no way to avoid this [13]?

**Review by michaelmcinerney — APPROVED** [review]
Awesome

File: proof/refine/ARM/Syscall_R.thy [hunk 4]
```diff
@@ -1522,72 +1522,181 @@ crunch sch_act_sane[wp]: cteDeleteOne sch_act_sane
    simp: crunch_simps unless_def
    rule: sch_act_sane_lift)
 
+lemma getCapReg_corres_gen:
+  "corres (\<lambda>x y. x = to_bl y) cur_tcb cur_tcb'
+          (get_cap_reg rg) (getCapReg rg)"
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:1527]: This and some of the following lemmas sort of feels like they should be earlier, but I guess if nothing else uses them then here is fine.

File: proof/refine/ARM/Syscall_R.thy [hunk 5]
```diff
@@ -1522,72 +1522,181 @@ crunch sch_act_sane[wp]: cteDeleteOne sch_act_sane
...
+  "\<lbrace> valid_objs' \<rbrace> lookupReply \<lbrace> valid_cap' \<rbrace>, -"
+  unfolding lookupReply_def getCapReg_def
+  apply (wpsimp wp: get_cap_wp hoare_vcg_imp_lift_R)
+       apply (rule hoare_FalseE_R)
+      apply wpsimp+
+  done
+
+lemma getBoundBNotification_corres: (* generalise *)
+  "corres (=) (ntfn_at nptr) (ntfn_at' nptr)
+          (get_ntfn_obj_ref ntfn_bound_tcb nptr) (liftM ntfnBoundTCB (getNotification nptr))"
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:1593]: Nitpick: is that comment meant to have a FIXME?

File: proof/refine/ARM/Syscall_R.thy (hunk 3 above)
Comment (corlewis) [reply proof/refine/ARM/Syscall_R.thy]: It looks like you managed to avoid it, nice!

File: proof/refine/ARM/Syscall_R.thy [hunk 6]
```diff
@@ -2039,17 +2171,24 @@ lemma handleYield_corres:
...
-  apply (wp ct_in_state_thread_state_lift'
-            rescheduleRequired_invs'
-            tcbSchedAppend_invs' | simp)+
-  apply (clarsimp simp add: invs'_def ct_in_state'_def sch_act_wf_weak cur_tcb'_def
-                   valid_pspace_valid_objs' valid_objs'_maxDomain tcb_in_cur_domain'_def)
-  sorry (*
-  apply (simp add:ct_active_runnable'[unfolded ct_in_state'_def])
-  done *)
+  by (wpsimp wp: updateSchedContext_invs' ct_in_state_thread_state_lift'
+    | strengthen live_sc'_ex_cap[OF invs_iflive'] valid_sc_strengthen[OF invs_valid_objs'])+
```
Comment (corlewis) [inline proof/refine/ARM/Syscall_R.thy:2191]: nitpick
```suggestion
  by (wpsimp wp: updateSchedContext_invs' ct_in_state_thread_state_lift'
      | strengthen live_sc'_ex_cap[OF invs_iflive'] valid_sc_strengthen[OF invs_valid_objs'])+
```

**Review by corlewis — APPROVED** [review]
Extra nice job cleaning up those preconditions!

**Pull request merged**